File: 3.x.x/crates-utopia/lock/src/file.rs
use std::fs::{File, OpenOptions};
use std::path::{Path, PathBuf};
use std::thread;
use std::time::{Duration, Instant};

use parking_lot::Mutex;

use crate::error::{Contention, LockError};
use crate::lock::Lock;

/// PHP `LOCK_SH`.
pub const LOCK_SH: i32 = 1;
/// PHP `LOCK_EX`.
pub const LOCK_EX: i32 = 2;

/// PHP `Utopia\Lock\File`.
#[derive(Debug)]
pub struct FileLock {
    path: PathBuf,
    exclusive: bool,
    handle: Mutex<Option<File>>,
}

impl FileLock {
    pub fn new(path: impl AsRef<Path>, mode: i32) -> Self {
        Self {
            path: path.as_ref().to_path_buf(),
            exclusive: mode != LOCK_SH,
            handle: Mutex::new(None),
        }
    }

    pub fn with_exclusive(path: impl AsRef<Path>) -> Self {
        Self::new(path, LOCK_EX)
    }

    fn open(&self) -> Result<(), LockError> {
        let mut slot = self.handle.lock();
        if slot.is_some() {
            return Ok(());
        }
        let directory = self.path.parent().unwrap_or_else(|| Path::new("."));
        if !directory.is_dir() {
            return Err(LockError::new(format!(
                "Lock file directory does not exist: {}",
                directory.display()
            )));
        }
        let file = OpenOptions::new()
            .create(true)
            .truncate(false)
            .read(true)
            .write(true)
            .open(&self.path)
            .map_err(|err| LockError::new(format!("Failed to open lock file: {err}")))?;
        *slot = Some(file);
        Ok(())
    }

    fn try_flock(&self) -> bool {
        if self.open().is_err() {
            return false;
        }
        let slot = self.handle.lock();
        let Some(file) = slot.as_ref() else {
            return false;
        };
        let result = if self.exclusive {
            fs2::FileExt::try_lock_exclusive(file)
        } else {
            fs2::FileExt::try_lock_shared(file)
        };
        result.is_ok()
    }
}

impl Lock for FileLock {
    fn acquire(&self, timeout: f64) -> bool {
        if timeout <= 0.0 {
            return self.try_acquire();
        }
        let deadline = Instant::now() + Duration::from_secs_f64(timeout);
        let mut delay = Duration::from_millis(10);
        loop {
            if self.try_acquire() {
                return true;
            }
            let remaining = deadline.saturating_duration_since(Instant::now());
            if remaining.is_zero() {
                return false;
            }
            thread::sleep(delay.min(remaining));
            delay = (delay * 2).min(Duration::from_millis(250));
        }
    }

    fn try_acquire(&self) -> bool {
        self.try_flock()
    }

    fn release(&self) {
        let mut slot = self.handle.lock();
        if let Some(file) = slot.take() {
            let _ = fs2::FileExt::unlock(&file);
            drop(file);
        }
    }

    fn contention(&self) -> Contention {
        Contention::new(format!(
            "Failed to acquire file lock on {} within timeout",
            self.path.display()
        ))
    }
}

File: packages/lock/src/File.php
<?php

declare(strict_types=1);

namespace Utopia\Lock;

use RuntimeException;
use Utopia\Lock\Exception\Contention;

final class File implements Lock
{
    /**
     * @var resource|null
     */
    private $handle;

    public function __construct(
        private readonly string $path,
        private readonly int $mode = LOCK_EX,
    ) {}

    #[\Override]
    public function acquire(float $timeout = 0.0): bool
    {
        $this->open();

        if ($timeout <= 0.0) {
            return $this->tryAcquire();
        }

        $deadline = microtime(true) + $timeout;
        $delay = 0.01;

        do {
            if ($this->tryAcquire()) {
                return true;
            }

            $remaining = $deadline - microtime(true);
            if ($remaining <= 0.0) {
                return false;
            }

            usleep((int) (min($delay, $remaining) * 1_000_000));
            $delay = min($delay * 2.0, 0.25);
        } while (true);
    }

    #[\Override]
    public function tryAcquire(): bool
    {
        $this->open();

        \assert($this->handle !== null);

        $wouldBlock = 0;

        /** @var int<0, 7> $operation */
        $operation = $this->mode | LOCK_NB;

        return flock($this->handle, $operation, $wouldBlock);
    }

    #[\Override]
    public function release(): void
    {
        if ($this->handle === null) {
            return;
        }

        flock($this->handle, LOCK_UN);
        fclose($this->handle);
        $this->handle = null;
    }

    #[\Override]
    public function withLock(callable $callback, float $timeout = 0.0): mixed
    {
        if (! $this->acquire($timeout)) {
            throw new Contention("Failed to acquire file lock on {$this->path} within timeout");
        }

        try {
            return $callback();
        } finally {
            $this->release();
        }
    }

    private function open(): void
    {
        if ($this->handle !== null) {
            return;
        }

        $directory = \dirname($this->path);
        if (! is_dir($directory)) {
            throw new Exception("Lock file directory does not exist: {$directory}");
        }
        if (! is_writable($directory) && ! file_exists($this->path)) {
            throw new Exception("Lock file directory is not writable: {$directory}");
        }

        $handle = fopen($this->path, 'c');
        if ($handle === false) {
            throw new RuntimeException("Failed to open lock file: {$this->path}");
        }

        $this->handle = $handle;
    }
}
